### DISCUSSION TITLE
STRSTR new instruction

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#70] [General]
I'm playing with René's substr proposal and added a new instruction STRSTR which includes the ITOS function. It's just a minimal improvement, but a least we save one instruction per character. Please let me know, if I should proceed I'll then push the new code. 

/* test appendchar */
options levelb

say 'test appendchar'
string='The quick brown fox jumps over the lazy dog'
substr=''
chr=''
offset=5
len=5
do i=offset to offset+len-1
   assembler do
     strstr chr,string,i
/*   itos chr */
     appendchar substr,chr
   end
end
say substr

**Comment by Peter-Jacob** [comment]
**Result is:**
C:\Users\PeterJ\CLionProjects\CREXX\cmake-build-debug\interpreter\rxvm.exe test
test appendchar
quick

Process finished with exit code 0

**Reply by Peter-Jacob** [reply]
hm, maybe we could even add an **appendstr substr,string,offset** instruction ...

**Comment by adesutherland** [comment]
I don't think the itos was needed anyway?

appendchar takes the in value (op2) and strchar sets the int value. Worth checking if this is right ...


What could be good is (building on your suggestion in the next message) is a kind of bulk copy

appendstr target, source, length - which copies length characters from the **current position** in source (if that makes sense?) and appends to target

A

**Reply by Peter-Jacob** [reply]
makes absolutely sense! The problem is 'source' doesn't have an offset, just a length. I wonder if we could use an implicit register, not part of the instruction to define it. IBM assembler does it for certain instructions MVCL for example, using odd/even register pairs.

**Reply by adesutherland** [reply]
Ah ... but the good news it does now :-) 

Each register has a strpos attribute - see setstrpos and getstrpos instructions

Also I just found this instruction - "substr", "op1 = op2[charpos]...op2[charpos+op3-1]" (which is unimplemented). But don't tell Rene!

**Reply by Peter-Jacob** [reply]
concerning itos: it is necessary, checked it.

**Reply by Peter-Jacob** [reply]
> Each register has a strpos attribute - see setstrpos and getstrpos instructions

this is an interesting proposal!

**Reply by adesutherland** [reply]
> concerning itos: it is necessary, checked it.
Mmm ... weird. I need to look at this.

**Comment by rvjansen** [comment]
Yes, I took the it’s out because it made no sense.

**Comment by rvjansen** [comment]
Hmm spell check needs to be switched off on this machine also

The ITOS seems not necessary. I wonder how you found it is.

**Reply by Peter-Jacob** [reply]
say 'test appendchar'
string='The quick brown fox jumps over the lazy dog'
substr=''
chr=''
offset=5
len=5
do i=offset to offset+len-1
   assembler do
     strchar chr,string,i
/*   itos chr */
     appendchar substr,chr
   end
end
say substr


without itos, it doesn't work.

**Reply by adesutherland** [reply]
It worked for me without itos - fyi this is the rxas produced. Weird.

```
/* REXX COMPILER PoC */

.globals=0

main()   .locals=7
   * Line 2: say 'test appendchar'
   say "test appendchar"
   * Line 3: string='The quick brown fox jumps over the lazy dog'
   load r1,"The quick brown fox jumps over the lazy dog"
   * Line 4: substr=''
   load r2,""
   * Line 5: chr=''
   load r3,""
   * Line 8: {DO}
   * Line 8: i=offset
   load r5,5
   copy r4,r5
   * Line 8: to offset+len-1
   load r6,9
l39:
   * Line 8: to offset+len-1
   igt r0,r4,r6
   brt l30,r0
   * Line 10: strchar chr,string,i
   strchar r3,r1,r4
   * Line 12: appendchar substr,chr
   appendchar r2,r3
   * Line 8: {Implicit "BY 1"}
   inc r4
   * Line 8: {DO-END}
   br l39
l30:
   * Line 15: say substr
   say r2
   ret
```

Minor - and doesn't change anything but line 5 chr='' should be chr = 0 as it is an int.

Also the compiler is generating some braindead code here:
```
 * Line 8: i=offset
   load r5,5
   copy r4,r5
```
I need to look at that!

A

**Reply by Peter-Jacob** [reply]
Sorry, my fault! I have overseen a change I made in STRCHAR for a pre-test of STRSTR. itos is not necessary for appendchr! This means we don't need an STRSTR instruction. So, go back to square 1 ;)

**Reply by adesutherland** [reply]
Thanks for letting us know! :-)

**Reply by adesutherland** [reply]
> Also the compiler is generating some braindead code here:
> 
>  * Line 8: i=offset
>    load r5,5
>    copy r4,r5
> 

Fixed

**Comment by rvjansen** [comment]
I have:

loop1:
          /* a2:int argument 2, start position */
          /* r10:int length of arg */
          /* r4:string argument1, string */
          strchar r13,r4,a2
          iadd a2,a2,1 /* increment loop index */
          igt r11,a2,r10
          br print
resume: 
          brt endloop1,r11 /* jump out of the loop */
          appendchar r14,r13
          br  loop1
endloop1:
          ret r14
          exit

Works fine, just still working on the variants with length and pad.

**Comment by rvjansen** [comment]
To update, this works, with and without length option, pad still tbd:

```
.globals=0

substr()  .expose=global.substr
    
main()    .locals=8
          load r1,"Test"
          load r2,"of Substr function"
          sconcat r2,r1,r2
          say r2

         /* the first example from the book, SUBSTR('abc',2) == 'bc' */
          load r3,2  /* there are two arguments*/
          load r4,""
          load r4,"abc"
          load r5,2 /* second argument, startpos */
          call r1,substr(),r3
          say r1

          /* the second example from the book, SUBSTR('abc',2,4) == 'bc  ' */
          load r3,3  /* there are three arguments*/
          load r4,""  
          load r4,"abcdefgh"
          load r5,3 /* second argument, startpos */
          load r6,6 /* third argument, length */   
          call r1,substr(),r3
          say r1

          /* the third example from the book, SUBSTR('abc',2,6,'.') == 'bc....' */
          load r3,4  /* there are four arguments*/
          load r4,""
          load r4,"abcdefghijklmnopqrstuvwxyz"
          load r5,1 /* second argument, startpos */
          load r6,26 /* third argument, length */
          load r7,"." /* fourth argument, pad */ 
          call r1,substr(),r3
          say r1
          
          exit
          
```

which tests the following function:

```
/* 
 * rexx SUBSTR bif in rxas. Uses length.rxas
 * rudimentary error checking
 */

          .globals=0
length()  .expose=global.length          
substr()  .locals=16 .expose=global.substr
          /* string    = arg1    */
          /* n (start) = arg2    */
          load r7," " /* the default pad character */
          load r6,0   /* the length field initialized to 0 */
          ilt r1,a0,2 /* less than 2 arguments */
          brt wrongargs,r1
          igt r1,a0,4 /* more than 4 arguments */
          brt wrongargs,r1
          ieq r1,a0,2
          brt doit,r1 /* we have two arguments */
          /* we have 3 or 4 arguments */
          move r6,a3 /* save length field in r6*/
          ieq r1,a0,4
          brf doit,r1 /* we do not have 4 arguments, no pad */
          move r7,a4
          br doit

wrongargs:
          /* we are here if there are not enough, or too many arguments */
          say "substr needs at least two and at most 4 arguments"
          exit
doit:
          /* we want to know the length of the string argument */
          load r3,1 /* there is one argument for the call to length() */
          move r4,a1 /* and it is the string in a1 */
          call r10,length(),r3 /* what is the length of the string arg */
          move r12,r10 /* the preliminary loop counter is the strlength */
          iadd r12,r12,1
          /* if r6 is nonzero, we have a length argument, a2:=r8+1 */
          igt r1,r6,0
          brf skip,r1
          move r12,r6 /* the final loop counter is the length argument */
          iadd r12,r12,a2
skip:    
          /* now loop characters from a1 from the start pos in a2 to (length) r1 */
          load r14,"" /* we are going to return this so initialize it */

loop1:
          /* a2:int argument 2, start position */
          /* r4:string argument1, string */   
          /* r10:int length of arg */
          /* r12: loop counter */
          /* r13: receive integer of r4[a2] */
          strchar r13,r4,a2
          iadd a2,a2,1 /* increment loop index */
          igt r11,a2,r12
          /*br print */
resume: 
          brt endloop1,r11 /* jump out of the loop */
          appendchar r14,r13
          br  loop1
endloop1:
          ret r14
          exit

/*
print:    itos a2
          itos r12
          sconcat r15,a2,r12
          say r15
          br resume
*/
```
uncomment to debug the length and index fields.

**Comment by rvjansen** [comment]
replaced some iadds with inc (on Peter's suggestion) and implemented pad. This runs the example from the Rexx book correctly.

```
/* 
 * rexx SUBSTR bif in rxas. Uses length.rxas
 * rudimentary error checking
 */
          .globals=0
length()  .expose=global.length          
substr()  .locals=18 .expose=global.substr
          /* string    = arg1    */
          /* n (start) = arg2    */
          load  r7," " /* the default pad character */
          load  r6,0   /* the length field initialized to 0 */
          ilt   r1,a0,2 /* less than 2 arguments */
          brt   wrongargs,r1
          igt   r1,a0,4 /* more than 4 arguments */
          brt   wrongargs,r1
          ieq   r1,a0,2
          brt   doit,r1 /* we have two arguments */
          /* we have 3 or 4 arguments */
          move  r6,a3 /* save length field in r6*/
          ieq   r1,a0,4
          brf   doit,r1 /* we do not have 4 arguments, no pad */
          move  r7,a4
          br    doit

wrongargs:
          /* we are here if there are not enough, or too many arguments */
          say "substr needs at least two and at most 4 arguments"
          exit
doit:
          /* we want to know the length of the string argument */
          load  r3,1 /* there is one argument for the call to length() */
          move  r4,a1 /* and it is the string in a1 */
          call  r10,length(),r3 /* what is the length of the string arg */
          load  r17,0
          inc   r17
          iadd  r17,r17,r10 /* because the next move clobbers r10 */
          move  r12,r10 /* the preliminary loop counter is the strlength */
          inc   r12
          /* if r6 is nonzero, we have a length argument, a2:=r8+1 */
          igt   r1,r6,0
          brf   skip,r1
          move  r12,r6 /* the final loop counter is the length argument */
          iadd  r12,r12,a2
skip:    
          /* now loop characters from a1 from the start pos in a2 to (length) r1 */
          load  r14,"" /* we are going to return this so initialize it */

loop:
          /* a2:int argument 2, start position */
          /* r4:string argument1, string */
          /* r7: the pad character */
          /* r17:int length of arg */
          /* r12: loop counter */
          /* r13: receive integer of r4[a2] */
          strchar r13,r4,a2
          inc   a2 /* increment loop index */
          igt   r11,a2,r12
          brt   endloop,r11 /* jump out of the loop */
          igt   r16,a2,r17
          brf   notpadded,r16 
          concat r14,r14,r7  /* we are past lengt, concat pad */
          br loop
notpadded:
          appendchar r14,r13 /* still adding part of string */
          br    loop
endloop:
          ret   r14
          exit

```

**Comment by rvjansen** [comment]
update: I have working versions & tests of LENGTH(), SUBSTR(), LEFT() and COPIES() now - will do RIGHT() tomorrow and will check those in in a fresh version of F004, following the kanban planning; all other functions are fair game.

**Reply by adesutherland** [reply]
BOOM :-)

**Comment by adesutherland** [comment]
I am working on REXX declaring and calling functions. Declaring is mostly done. I still need to update RETURN and add validation that functions always return when they should, including returning nothing.

For calling, parsing, validation, type checking and the ability to handle "a" and "g" as well as "r" registers is done. I have got to the point of the algorithm to assign registers for the function call reasonably efficiently (it needs a bunch of registers in order, like r5,r6,r7 to pass 2 arguments), almost there. And them just emitting the code which is actually not a big job.

In sum, we should be calling some of your BIFs form REXX over the weekend (fingers crossed).

**Reply by rvjansen** [reply]
"functions always return when they should, including returning nothing."

Scritti Politti:
There's nothing I wouldn't do
Including doing nothing

Would be great if we can come to the point of calling BIF's - much earlier than expected.
